### ISSUE TITLE
Implement the TLS extension

### ISSUE THREAD

**Issue by pgjones** [pgjones/hypercorn#62]
See [spec](https://asgi.readthedocs.io/en/latest/specs/tls.html), [Uvicorn](https://github.com/encode/uvicorn/pull/1119), and [nonecorn](https://github.com/nonebot/nonecorn/commit/836f3ce9a2fe01a0439b896bdfe67de303c03727). 

Looks as if the specification is not implementable as currently defined, therefore I will wait to see if the specification is updated.

**Comment by synodriver** [comment]
Yes, I'm willing to make a pull request for that since nonecorn is forked from hypercorn and should be 100% compatible with upstream. But since nonecorn used to trace gitlab's branch of hypercorn, it seems that github is prevent me from making that pull request    :)

**Comment by pgjones** [comment]
Does the nonecorn code implements the full specification? When I looked it was similar to the Uvicorn PR in that not everything could be implemented. If so I think we need to change the ASGI specification before we commit to implementing it.

**Comment by synodriver** [comment]
Indeed, nonecorn, just like the uvicorn PR, did not implement the full specification, it left some keys to None, and according to the [spec](https://asgi.readthedocs.io/en/latest/specs/tls.html), those keys are optional. So which part do you think can't be implemented? I'll dig deeper and have a try.

Besides, I only ported it to asyncio, trio is not implemented yet, therefore more efforts is needed.

**Comment by commonism** [comment]
Required for OpenAPI mutualTLS authentication - e.g. FastAPI/starlette/hypercorn.

**Comment by commonism** [comment]
I'd start off with nonecorn instead - it does asgi tls ext already.
https://github.com/nonebot/nonecorn/commit/836f3ce9a2fe01a0439b896bdfe67de303c03727

**Comment by pquentin** [comment]
Yes but it suffers from a few issues: it does not support Trio (which is what urllib3 uses), it's a bit hacky (various functions accept `ssl` *and* `tls` parameters), and to my knowledge does not have tests for the TLS extension. It also does not support `alpn_protocols`, so urllib3 needed a temporary fork anyway. More generally, I don't like the idea of having a long-term fork for things that Hypercorn itself wants. Working together would produce a better outcome.

**Comment by lschneidpro** [comment]
any update ?

**Comment by commonism** [comment]
yes - https://github.com/encode/uvicorn/pull/1119

**Comment by grydz** [comment]
It would be great to have a partial support of the ASGI TLS Extension, but because it's definitively not the role of the ASGI web server to extract information from the client certificate,  I suggest to let the `client_cert_name` empty.

Note that certificate parsing can be done easily on the application side with external libraries such as [cryptography](https://cryptography.io/en/latest/):

```python
if "tls" in request.scope["extensions"]:
    if "client_cert_chain" in request.scope["extensions"]["tls"]:
        client_cert, *_ = request.scope["extensions"]["tls"]["client_cert_chain"]
        cert = x509.load_pem_x509_certificate(client_cert.encode("utf-8"))
        name_attr, *_ = cert.subject.get_attributes_for_oid(
            x509.NameOID.COMMON_NAME
        )
        cn = name_attr.value
```


Based on https://github.com/encode/uvicorn/pull/1119 and @pquentin's [branch](https://github.com/urllib3/hypercorn/tree/tls-extension), is something like https://github.com/grydz/hypercorn/compare/main...asgi-tls-ext could be acceptable?

**Comment by rdejana** [comment]
Any progress on this?
